engine/access/rpc/backend/connection_factory.go

```diff
@@ -81,14 +83,16 @@ func (cf *ConnectionFactoryImpl) retrieveConnection(grpcAddress string, timeout
 		}
 	}
 	if conn == nil || conn.GetState() != connectivity.Ready {
-		// This line ensures that when a connection is renewed, the previously cached connection is evicted and closed
-		cf.ConnectionsCache.Remove(grpcAddress)
 		var err error
 		conn, err = cf.createConnection(grpcAddress, timeout)
 		if err != nil {
 			return nil, err
 		}
+		cf.lock.Lock()
```

Comment:
isn't the connection cache concurrency safe? Would this lock slow down ALL connection request?

Reply:
It's true that the connection cache is concurrency safe but that's only true for each independent operation. Our concern was that there may exist a race condition in updating the cache where it might look something like this:

#1 - cf.ConnectionsCache.Remove(grpcAddress)
#2 - cf.ConnectionsCache.Remove(grpcAddress)
#1 - cf.ConnectionsCache.Add(grpcAddress, conn)
#2 - cf.ConnectionsCache.Add(grpcAddress, conn)

Reply:
there's still a race condition here, except it will result in connection thrashing instead of connection leaking. 

1. Thread A get's the connection
2. Thread B get's the connection
3. Thread A determines the connection isn't good anymore and takes the lock
4. Thread B determines the connection isn't good anymore and waits for the lock
5. Thread A removes the old connection, adds a new one, then releases the lock
6. Thread B gets the lock, then removes the new connection and adds a different new one

In this case, Thread A's connection would be closed prematurely. I'm not 100% what happens here, but I'm guessing the connection gets reset and aborted early. We should find out.

To fix that we'll need to move the locking to before the initial get. As Vishal noted this will likely cause lock contention, so we may need to reduce the scope of the lock to a single address.

Reply:
> To fix that we'll need to move the locking to before the initial get. 

Or, we think it's not very common for the connection to be "not good", we could add another `Get` inside the lock, so just below this line, and do another check. If it's good, use, if it's not, then remove and add the good connection.

It just depends how often connections need to get re-cycled.